content/blog/2019-09-25-RecordTypes.md

```diff
@@ -0,0 +1,131 @@
+
+
+We chose this format because using the dot operator to access fields is very common in modern programming languages.
+
+### *With* Syntax
+While immutable data structures allow you to more easily reason about how values flow through your program, these currently immutable records are cumbersome to change. 
+For example, to change one field, you must recreate the entire record. While we do want a new record to be created, copying variables between two records is overly tedious. 
+As such, we decided to implement a with syntax, similar to OCaml in which the user just needs to specify a record name as the base, and then the fields desired to be changed, along with the new value. This maintains immutability, without so much tedium.
+
+We use the following syntax:
+
+    <new record>: <record type> = 
+        <old record> with {<field1 name>: <field1 value>; <field2 name>: <field2 value> … };
+
+Where:
+`<new record>` is the name of the new record
+`<record type>` is the same type as `<old record>`
+`with` is a new keyword
+`<field# name>` must match with a field in `<type>`
+`<field# value>` must be a variable with a type that matches its field name.
+Within the braces, the user may specify 0 to n field name and value pairs, where n is the total number of fields in the record type.
+
+This syntax was designed to have a similar format as record instantiation. 
+
+### Immutability
+Immutable data structures are easier to reason about when developing compiler optimizations. However, forcing developers to recreate the entire record every time they need to change a value poses usability challenges. We decided on sticking to immutability in part for optimizations' sake. The ability to make guarantees that a value will not change lends itself well to constant-folding, even after a function call that takes in your record as an argument. 
+
+### Evaluation
+We set out to implement record types and we successfully implemented immutable nominal record types. Record types do not have any restrictions except that two fields cannot have the same name and that a record type cannot be recursive. Besides that, everything is allowed. Record types play nicely with existing types and new syntactical additions follow previous precedents set by Bril or otherwise follow precedents set by other languages that have record types. Record declaration, instantiation, and access are all (we think) intuitive and straightforward, and this means our record types provide good value as an addition to the Bril language. 
+
+We found that creating new records was a tedious process if the record was large, so we implemented *with* statements in addition to the features mentioned above for situations where one wanted to duplicate a record with a few changes. It should be noted that it is bad form to use a with statement with no fields because that would be identical to referencing the old record with `… = id oldRecordName.`
```

Comment:
```suggestion
We found that creating new records was a tedious process if the record was large, so we implemented *with* statements in addition to the features mentioned above for situations where one wanted to duplicate a record with a few changes. It should be noted that it is bad form to use a with statement with no fields because that would be identical to referencing the old record with `… = id oldRecordName`.
```